staging/src/k8s.io/client-go/dynamic/simple.go

```diff
@@ -303,27 +306,97 @@ func (c *dynamicResourceClient) Get(name string, opts metav1.GetOptions, subreso
 }
 
 func (c *dynamicResourceClient) List(opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
-	result := c.getRestClient().Get().AbsPath(c.makeURLSegments("")...).SpecificallyVersionedParams(&opts, dynamicParameterCodec, versionV1).Do()
-	if err := result.Error(); err != nil {
-		return nil, err
-	}
-	retBytes, err := result.Raw()
-	if err != nil {
-		return nil, err
-	}
-	uncastObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, retBytes)
-	if err != nil {
-		return nil, err
-	}
-	if list, ok := uncastObj.(*unstructured.UnstructuredList); ok {
-		return list, nil
-	}
+	// When resource version is not empty, it reads from api server local cache
+	// Need to check all api server partitions
+	searchClients := make([]*rest.RESTClient, 1)
+	if opts.ResourceVersion != "" && len(c.client.clients) > 1 {
+		searchClients = c.client.clients
+	} else {
+		searchClients[0] = c.getRestClient()
+	}
+
+	var listLock sync.Mutex
+
+	var wg sync.WaitGroup
+	results := make(map[int]*unstructured.UnstructuredList)
+	errs := make(map[int]error)
+	for i, client := range searchClients {
+		wg.Add(1)
+
+		go func(ci rest.Interface, opts metav1.ListOptions, lock *sync.Mutex, pos int, resultMap map[int]*unstructured.UnstructuredList, errMap map[int]error) {
+			defer wg.Done()
+
+			r := ci.Get().AbsPath(c.makeURLSegments("")...).SpecificallyVersionedParams(&opts, dynamicParameterCodec, versionV1).Do()
+			if err := r.Error(); err != nil {
+				setResult(resultMap, errMap, lock, pos, nil, err)
+				return
+			}
+
+			retBytes, err := r.Raw()
+			if err != nil {
+				setResult(resultMap, errMap, lock, pos, nil, err)
+				return
+			}
+
+			uncastObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, retBytes)
+			if err != nil {
+				setResult(resultMap, errMap, lock, pos, nil, err)
+				return
+			}
+
+			if list, ok := uncastObj.(*unstructured.UnstructuredList); ok {
+				setResult(resultMap, errMap, lock, pos, list, nil)
+				return
+			}
+
+			list, err := uncastObj.(*unstructured.Unstructured).ToList()
+			if err != nil {
+				setResult(resultMap, errMap, lock, pos, nil, err)
+				return
+			} else {
+				setResult(resultMap, errMap, lock, pos, list, nil)
+			}
+
+		}(client, opts, &listLock, i, results, errs)
+	}
+	wg.Wait()
```

Comment:
Nit:  A shared channel + select is probably is better here, as we can start merging results without waiting for everyone to finish.  But I guess the cost will mostly be networking and server side processing, so the actual difference should be very minor here.

Reply:
The time saved for starting merge earlier is not significant. Would like to use more straight forward code to prevent potential bugs.